oeps/oep-0002-bp-repo-metadata.rst

```diff
@@ -29,74 +30,119 @@ Proposes a standard format for repository metadata for Open edX projects.
 Motivation
 ==========
 
-One of the primary goals for Best Practice OEPs is to standardize techniques
-and tools across Open edX repositories. However, without
-tooling, it will be difficult to move all of the repositories in the same
-direction (and hard to validate which ones have been updated). This OEP exists
-to fill that gap.  By providing a standard format for metadata about accordance
-to Best Practice OEPs, tools can be developed in the future to report on the
-current status of all of the Open edX repositories, and how well they follow
-the Best Practices listed in the OEPs.
+One of the primary goals for Best Practice OEPs is to standardize techniques and tools across Open edX repositories. However, without tooling, it will be difficult to move all of the repositories in the same direction (and hard to validate which ones have been updated). This OEP exists to fill that gap.  By providing a standard format for metadata about accordance to Best Practice OEPs, tools can be developed in the future to report on the current status of all of the Open edX repositories, and how well they follow the Best Practices listed in the OEPs.
 
-Bear in mind that particular classes of Open edX repositories may have
-different Best Practices. The practices that apply to an XBlock may not be the
-same as those that apply to a standalone Django service or to a Django app.
+Bear in mind that particular classes of Open edX repositories may have different Best Practices. The practices that apply to an XBlock may not be the same as those that apply to a standalone Django service or to a Django app.
 
 Specification
 =============
 
+Keys
+----
+
 Each repo will include a file ``openedx.yaml``, with the following keys:
 
-``owner``: string (required unless ``archived`` is ``True``)
-    The GitHub user responsible for the repository.
+``owner_type``: string (required)
```

Comment:
Agree, I think we should just make it clear that one of those two fields is required.

Reply:
Interesting... I tend to appreciate explicitness over implicitness.

Advantages of being explicit of `owner_type`:
1. **Writers** Having the writer/maintainer of the `openedx.yaml` file be explicit of the ownership model for that repo forces them to think about its implications.

2. **Readers** Downstream tools are simplified since they don't have to have `if-then` statements and understand precedence order in case a repo had multiple `owning` fields entered.

3. **Updates** In case we ever added a new ownership model or allowed for fallbacks (both `owning_team` and `owning_repo` are specified but `owning_team` takes primary precedence, for example), we can do so more easily. All Readers will know that a new ownership model has been introduced since there would be a new value for `owner_type`.

Disadvantages of being explicit of `owner_type`:
1. One additional field to enter per repo.

Reply:
Maybe owner should just be a dict `type` and `value` attributes that are both required?

This way we can be explicit without repetition and there is one field that holds the answer for who is the owner.

Reply:
Sure. We can do that.

Reply:
Suggestion on the name of the `owner` dict?  It may be best if it's a different name from the existing `owner` name, right?  So we can automate finding which repos still need to be updated.

Reply:
PR is updated with `owner` as a dict.  Please review.

Reply:
Seems like we still have "type" which must be "team" or "repo", and then there is a "team" key or a "repo" key.  This redundancy just seems like more opportunity for error without adding any more information.

Reply:
I prefer `type` and `value`, which removes the redundancy. Unfortunately YAML doesn't have sum types :P

Reply:
I'm not sure what you mean by "doesn't have."  Wouldn't this be sufficient:
```
owner:
  repo: edx/edx-platform
```
or:
```
owner:
  team: edx/arch-bom
```

Reply:
It's reasonable, but doesn't preclude:
```yaml
owner:
  repo: edx/edx-platform
  team: edx/arch-bom
```
A more "sum type" -like constructor would look something like:
```yaml
owner: repo(edx/edx-platform)
```
Of course, YAML isn't typed anyway, so anyone can write anything they want. So we just want something that is maximally readable and minimally ambiguous.

I'm fine with both your suggestion and @feanil 's suggestion, which was:
```yaml
owner:
  type: repo
  value: edx/edx-platform
```

Reply:
The spec could preclude both repo and team by saying, "Don't provide both repo and team."  Ultimately it's the tooling that's going to enforce any of these rules.

Reply:
@kdmccormick and @nedbat Can you explain how the other 2 suggestions could support the 3rd goal I listed above?

```
3. Updates. In case we ever added a new ownership model or allowed for fallbacks (both owning_team and owning_repo are specified but owning_team takes primary precedence, for example), we can do so more easily. 
```

With the `type` and `value` option, wouldn't value then have to be a `dict`?